brilift/src/translator.rs

```diff
@@ -0,0 +1,1104 @@
+    // `cranelift_jit` does not yet support PIC on AArch64:
+    // https://github.com/bytecodealliance/wasmtime/issues/2735
+    // The default initialization path for `JITBuilder` is hard-coded to use PIC, so we manually
+    // disable it here. Once this is fully supported in `cranelift_jit`, we can switch to the
+    // generic versin below unconditionally.
+    #[cfg(target_arch = "aarch64")]
+    fn jit_builder() -> JITBuilder {
+        let mut flag_builder = settings::builder();
+        flag_builder.set("use_colocated_libcalls", "false").unwrap();
+        flag_builder.set("is_pic", "false").unwrap(); // PIC unsupported on ARM.
+        let isa_builder = cranelift_native::builder().unwrap();
+        let isa = isa_builder
+            .finish(settings::Flags::new(flag_builder))
+            .unwrap();
+        JITBuilder::with_isa(isa, cranelift_module::default_libcall_names())
+    }
+
+    // The normal way to set up a JIT builder.
+    #[cfg(not(target_arch = "aarch64"))]
+    pub fn jit_builder() -> JITBuilder {
+        JITBuilder::new(cranelift_module::default_libcall_names()).unwrap()
+    }
+
+    pub fn new() -> Self {
+        // Set up the JIT.
+        let mut builder = Self::jit_builder();
+
+        // Provide runtime functions.
+        enum_map! {
+            rtfunc => {
+                let f: RTFunc = rtfunc;
+                builder.symbol(f.name(), f.rt_impl());
+            }
+        };
+
+        let mut module = JITModule::new(builder);
+
+        Self {
+            rt_funcs: Self::declare_rt(&mut module),
+            context: module.make_context(),
+            module,
+            funcs: HashMap::new(),
+        }
+    }
+
+    /// Obtain an entry-point code pointer. The pointer remains valid as long as the translator
+    /// itself (and therefore the `JITModule`) lives.
+    fn get_func_ptr(&mut self, func_id: cranelift_module::FuncId) -> *const u8 {
+        self.module.clear_context(&mut self.context);
+        self.module.finalize_definitions().unwrap();
+
+        self.module.get_finalized_function(func_id)
+    }
+
+    fn val_ptrs(vals: &[bril::Literal]) -> Vec<*const u8> {
+        vals.iter()
+            .map(|lit| match lit {
+                bril::Literal::Int(i) => i as *const i64 as *const u8,
+                bril::Literal::Bool(b) => b as *const bool as *const u8,
+                bril::Literal::Float(f) => f as *const f64 as *const u8,
+                bril::Literal::Char(c) => c as *const char as *const u8,
+            })
+            .collect()
+    }
+
+    /// Run a JITted wrapper function.
+    pub unsafe fn run(&mut self, func_id: cranelift_module::FuncId, args: &[bril::Literal]) {
+        let func_ptr = self.get_func_ptr(func_id);
+        let arg_ptrs = Self::val_ptrs(args);
+        let func = mem::transmute::<_, fn(*const *const u8) -> ()>(func_ptr);
+        func(arg_ptrs.as_ptr());
+    }
+}
+
+#[derive(FromArgs)]
+#[argh(description = "Bril compiler")]
+pub struct Args {
```

Comment:
Since the CLI args don't seem to be used here (only in `main`), perhaps this struct should move there?